src/server.h

```diff
@@ -766,10 +768,21 @@ typedef struct redisDb {
     list *defrag_later;         /* List of key names to attempt to defrag one by one, gradually. */
 } redisDb;
 
-/* Declare database backup that include redis main DBs and slots to keys map.
- * Definition is in db.c. We can't define it here since we define CLUSTER_SLOTS
- * in cluster.h. */
-typedef struct dbBackup dbBackup;
+/* State for the Slot to Key API, for a single slot. The keys in the same slot
+ * are linked together using dictEntry metadata. See also "Slot to Key API" in
+ * cluster.c. */
+struct clusterSlotToKeys {
+    uint64_t count;             /* Number of keys in the slot. */
+    dictEntry *head;            /* The first key-value entry in the slot. */
+};
+typedef struct clusterSlotToKeys clusterSlotsToKeysData[CLUSTER_SLOTS];
+
+/* Declare temporary database that include redis main DBs and slots to keys map.
+ * Used during diskless replication to store new data without putting server in LOADING state. */
+typedef struct tempDb {
+    redisDb *dbarray;
+    clusterSlotsToKeysData slots_to_keys;
+} tempDb;
```

Comment:
To keep the encapsulation of slot-to-key in cluster.c, we can make clusterSlotsToKeysData an opaque type. It means that it can only be accessed as a pointer. Functions in cluster.c need to be called to do anything with it.

Here in server.h, just declare it as a struct without declaring the fields.

```C
typedef struct clusterSlotsToKeysData clusterSlotsToKeysData; // <--- size of struct unknown here

typedef struct tempDb {
    redisDb *dbarray;
    clusterSlotsToKeysData *slots_to_keys; // <---- pointer
} tempDb;
```

The rest can be in cluster.c and cluster.h. (Actually encapsulation is not that great in Redis. We have too much stuff in the header files. Things like CLUSTER_SLOTS could actually be defined in cluster.c instead since it's only used there, but let's not solve that problem in this PR.)

```C
// cluster.h
#define CLUSTER_SLOTS 16384

typedef struct clusterSlotToKeys {
    uint64_t count;             /* Number of keys in the slot. */
    dictEntry *head;            /* The first key-value entry in the slot. */
} clusterSlotToKeys;

struct clusterSlotsToKeysData {
    clusterSlotToKeys by_slot[CLUSTER_SLOTS];
}

void slotToKeySwap(clusterSlotsToKeysData **slots_to_keys);

// cluster.c
void slotToKeySwap(clusterSlotsToKeysData **slots_to_keys) {
    clusterSlotsToKeysData *tmp = server.cluster.slots_to_keys;
    server.cluster.slots_to_keys = *slots_to_keys;
    *slots_to_keys = tmp;
}
```

I think `server.cluster.slots_to_keys` can also be a pointer to an allocated structure. It makes it possible to just swap the pointers without using memcpy.

Does this make sense?

Reply:
Thanks a lot for the tips @zuiderkwast
Do you mind adding a commit with your suggestion? (Or maybe a separated PR focused on organizing these types after this is merged as this is probably happening soon?)

Reply:
I can maybe make the opaque type preparations before this PR is merged. I prefer that code isn't moved back and forth too much in unstable.

Reply:
Here I made clusterSlotToKeyMapping opaque. I also made tempDb opaque (implementation fully in db.c). It's based on your branch with one commit added. https://github.com/zuiderkwast/redis/tree/feature/draft-use-tempdb-swapdb/opaque-types

`./runtest-cluster` fails, but it fails also with your branch for me. Did you run it?

Reply:
Oh, I always thought these were succeeded on PR build.
I just debugged and found 2 things:
It crashes on this block of initTempDb():
```
    /* Init cluster slots to keys map if enable cluster. */
    if (server.cluster_enabled) {
        memset(&tempDb->slots_to_keys, 0, sizeof(tempDb->slots_to_keys)); // <- crashing line
    }
```
Removing it and, as it should be, adjusting the test to watch for async_loading instead of loading will make it pass.
And of course, the test needs to be renamed to fit the changes made.

Considering it's an area you have previous experience, any suggestion?
This block is inspired on your:
```
/* Empty the slots-keys map of Redis Cluster. */
void slotToKeyFlush(void) {
    memset(&server.cluster->slots_to_keys, 0,
           sizeof(server.cluster->slots_to_keys));
}
```
But I added to ensure a clean tempDb->slots_to_keys on the when we initialize it prior to loading data on it.

Reply:
The slot-to-key data is supposed to match what's in server.db[0]. If the db is cleared the slot-to-key data should also be cleared. There is nothing complicated about it.

If you fixed it, then I guess you can add my commit (cherry-pick) and fix it the same way?

Reply:
I don get it. I haven't fixed... I don't understand why the current code fails still ;)

Reply:
I don't know either. I'm not familiar with async load. :)

Reply:
It's late here. I will look tomorrow.

Reply:
@oranagra @zuiderkwast 
Se last commit "Fix slots to keys handling on swapdb"
Correct me if I'm wrong but I don't think we actually needed to store it in the tempDb.
Previously it was flushed during backup, now I flush right before swapDb.
It the replication fails, it's still intact.
The cluster test also won't crash with this change anymore.

Reply:
Wait a second.

When we do async load, do we load the dump into tempDb and then swap? Is this how it works? Doesn't it mean that we need one slot-to-key mapping for the tempDB and another one for the main DB?

We can only clear the slot-to-key mapping when we clear the database. If we swap with a non-empty database, we need a matching slot-to-key mapping. If we clear the slot-to-key mapping when the database is not empty, it is easy to create a test case which crashes: Delete all keys. There is an assert in slotToKeyDelEntry which will fail.

Reply:
Having a bad feeling about the complexity of this whole change now =)
So while doing the replication, the calls to slotToKeyAddEntry should apply to a tempDb slots_to_keys, not to server.cluster->slots_to_keys as it has been in this PR so far? If that's the case, we need to send a pointer to the methods that change server.cluster->slots_to_keys to work on either the temp or server.cluster according to the situation.
Makes sense @oranagra?

Reply:
yes. sorry, i didn't closely follow the discussion about slots_to_keys.
as long as we keep serving from server.db we need the server.cluster->slots_to_keys to remain the old one.
during async loading, we also need dbAdd to update the new temporary slots_to_keys.
then if we discard the temp db, we discard the temp slots_to_keys, and if we swap the db to be the primary db, we need to copy the tmp slots_to_keys over the primary one.

Reply:
the last committed changes don't look correct yet.

p.s. i don't mind the fact that CLUSTER_SLOTS is moved to server.h, and maybe slots_to_keys being moved to be part of the database.
it should only be allocated when cluster is enabled, but i think trying to isolate that part in cluster.c/h and make redis unaware of how many hash slots there are is not worth it.
if @zuiderkwast has another idea i don't mind trying (didn't follow that part of the discussion).

p.p.s the cluster tests are indeed not executed on the PR CI (they're too slow), and also the only existing test for that may need some improvement to properly catch errors around this change, please look into that.

Reply:
Now hopefully that's what we need, please check again.
And thanks for the patience.

Reply:
The commit I made above makes slot-to-key mapping an opaque type. The isolation is good IMO. It's better in cluster.c because the slot hashing and everything else for cluster is there and does not need to be visible outside it.

It's easy to extend those functions to pass a SlotToKeyMapping* to slotToKeyAddEntry() and the other functions, as you said @eduardobr. Let's do that! :) That's all we need AFACT.

Reply:
@zuiderkwast as i said, i don't mind to expose this part of the cluster code in server.h (it's not as complicated as the cluster bus gossip or anything like that), it it's just a trivial listing of keys per slot,
i also don't have an objection to the opaque mechanism you created, which seems nice.
but the problem is still that unlike the handling of `dbArray` argument to `rdbLoadRio` and `db` argument to `dbAddRDBLoad`, which is able to make sure we add data into the temp db, while commands are serving from another (the one in the `server` struct), this mechanism isn't handling that yet. right?

I think that in theory, an opaque `slots_to_keys` pointer should be stored inside the redisDb struct, and if we do that, it solved the above problem (since the `db` pointer is passed around).
till now the cluster code got away from that problem because unlike non-clustered redis, redis-cluster only supports one database. but now, we actually have two.

p.s. we do have some future plan we have, to drop support for Sentinel, by allowing non-sharded clusters with multiple databases and voting replicas. in that case cluster will support multiple databases, but will actually not need the slot-to-key mapping.